// @lay2/pw-core/build/main/providers/eos-provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EosProvider = void 0;
const provider_1 = require("./provider");
const models_1 = require("../models");
const core_1 = __importDefault(require("@scatterjs/core"));
const eosjs2_1 = __importDefault(require("@scatterjs/eosjs2"));
const eosjs_ecc_1 = __importDefault(require("eosjs-ecc"));
const ckb_js_toolkit_1 = require("../ckb-js-toolkit");
const hashers_1 = require("../hashers");
const utils_1 = require("../utils");
const axios_1 = __importDefault(require("axios"));
class EosProvider extends provider_1.Provider {
    constructor(eosNetwork, onAddressChanged) {
        super(provider_1.Platform.eos);
        this.eosNetwork = eosNetwork;
        this.onAddressChanged = onAddressChanged;
    }
    async init() {
        console.log('[eos-provider] try scatterjs');
        const network = core_1.default.Network.fromJson(this.eosNetwork);
        core_1.default.plugins(new eosjs2_1.default());
        const connected = await core_1.default.scatter.connect('pwcore-app', {
            network,
        });
        if (!connected) {
            throw new Error('eos not connected');
        }
        console.log('[eos-provider] scatter connected');
        const scatter = core_1.default.scatter;
        await scatter.login();
        console.log('[eos-provider] scatter login');
        const account = scatter.identity.accounts.find((x) => x.blockchain === 'eos');
        const { name } = account;
        console.log('eos account', name);
        const lockArgs = await this.getCKBLockArgsForEosAccount(name);
        console.log('eos lock args', lockArgs);
        this.address = new models_1.Address(name, models_1.AddressType.eos, lockArgs);
        window.scatter = scatter;
        return this;
    }
    async getEosPublicKey(account) {
        const network = core_1.default.Network.fromJson(this.eosNetwork);
        const baseUrl = network.fullhost();
        const res = await axios_1.default.post(`${baseUrl}/v1/chain/get_account`, {
            account_name: account,
        });
        const data = res.data;
        const pubkey = data.permissions[0].required_auth.keys[0].key;
        return pubkey;
    }
    async getCKBLockArgsForEosAccount(account) {
        const pubkey = await this.getEosPublicKey(account);
        const publicKeyHex = eosjs_ecc_1.default.PublicKey(pubkey).toUncompressed().toHex();
        const publicHash = new hashers_1.Keccak256Hasher()
            .hash(new ckb_js_toolkit_1.Reader(`0x${publicKeyHex.slice(2)}`))
            .serializeJson();
        const address = '0x' + publicHash.slice(-40);
        return address;
    }
    processEosHash(a) {
        let str = a.replace('0x', '');
        str = utils_1.spliceStr(str, 12, 0, ' ');
        str = utils_1.spliceStr(str, 12 * 2 + 1, 0, ' ');
        str = utils_1.spliceStr(str, 12 * 3 + 2, 0, ' ');
        str = utils_1.spliceStr(str, 12 * 4 + 3, 0, ' ');
        str = utils_1.spliceStr(str, 12 * 5 + 4, 0, ' ');
        return str;
    }
    async sign(message) {
        const pubkey = await this.getEosPublicKey(this.address.addressString);
        const sig = await window.scatter.getArbitrarySignature(pubkey, this.processEosHash(message));
        const sigHex = eosjs_ecc_1.default.Signature.from(sig).toHex().replace('0x', '');
        let v = Number.parseInt(sigHex.slice(0, 2), 16);
        if (v >= 27)
            v = (v - 27) % 4;
        const result = '0x' +
            this.platform.toString(16).padStart(2, '0') +
            sigHex.slice(2) +
            v.toString(16).padStart(2, '0');
        return result;
    }
    close() {
        throw new Error('Method not implemented.');
    }
}
exports.EosProvider = EosProvider;
