// @lay2/pw-core/build/main/providers/tron-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TronProvider = void 0;
const provider_1 = require("./provider");
const models_1 = require("../models");
class TronProvider extends provider_1.Provider {
    constructor(onAddressChanged) {
        super(provider_1.Platform.tron);
        this.onAddressChanged = onAddressChanged;
    }
    async init() {
        if (!!window.tronWeb) {
            console.log('[tron-provider] try window.tronWeb');
            if (!window.tronWeb.defaultAddress ||
                !window.tronWeb.defaultAddress.base58) {
                throw new Error('get tron address failed, please switch to tron account');
            }
            this.address = new models_1.Address(window.tronWeb.defaultAddress.base58, models_1.AddressType.tron);
            return this;
        }
        else {
            throw new Error('window.tronWeb is undefined, Tron environment is required.');
        }
    }
    async sign(message) {
        let result = await window.tronWeb.trx.sign(message);
        let v = Number.parseInt(result.slice(-2), 16);
        if (v >= 27)
            v -= 27;
        result =
            '0x' +
                this.platform.toString(16).padStart(2, '0') +
                result.slice(2, -2) +
                v.toString(16).padStart(2, '0');
        return result;
    }
    close() {
        throw new Error('Method not implemented.');
    }
}
exports.TronProvider = TronProvider;
