// @lay2/pw-core/build/main/providers/web3modal-provider.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Web3ModalProvider = void 0;
const provider_1 = require("./provider");
const models_1 = require("../models");
const ethereum_ens_1 = __importDefault(require("ethereum-ens"));
const utils_1 = require("../utils");
const core_1 = __importStar(require("../core"));
class Web3ModalProvider extends provider_1.Provider {
    constructor(web3, onAddressChanged) {
        super(provider_1.Platform.eth);
        this.web3 = web3;
        this.onAddressChanged = onAddressChanged;
    }
    async init() {
        const accounts = await this.web3.eth.getAccounts();
        if (!utils_1.verifyEthAddress(accounts[0])) {
            throw new Error('get ethereum address failed');
        }
        this.address = new models_1.Address(accounts[0], models_1.AddressType.eth);
        if (this.web3.currentProvider.on) {
            this.web3.currentProvider.on('accountsChanged', async (newAccounts) => {
                this.address = new models_1.Address(newAccounts[0], models_1.AddressType.eth);
                if (this.onAddressChanged) {
                    this.onAddressChanged(this.address);
                }
            });
        }
        return this;
    }
    async ensResolver(ens) {
        try {
            return await new ethereum_ens_1.default(this.web3.currentProvider).resolver(ens).addr();
        }
        catch (e) {
            return 'Unknown ENS Name';
        }
    }
    async sign(message) {
        let result = await this.web3.eth.personal.sign(message, this.address.addressString, null);
        let v = Number.parseInt(result.slice(-2), 16);
        if (v >= 27)
            v -= 27;
        result =
            '0x' +
                (core_1.default.chainId === core_1.ChainID.ckb
                    ? ''
                    : this.platform.toString(16).padStart(2, '0')) +
                result.slice(2, -2) +
                v.toString(16).padStart(2, '0');
        return result;
    }
    async close() {
        if (this.web3 &&
            this.web3.currentProvider &&
            this.web3.currentProvider.close) {
            await this.web3.currentProvider.close();
        }
    }
}
exports.Web3ModalProvider = Web3ModalProvider;
