// @lay2/pw-core/build/module/models/script.js
import { Address, AddressType, getDefaultPrefix } from './address';
import { generateCkbAddressString } from '../utils';
import { validators, transformers, normalizers } from '../ckb-js-toolkit';
import { SerializeScript } from '../ckb-lumos/core';
import { Blake2bHasher } from '../hashers';
import { NervosAddressVersion } from '../helpers/address';
export class Script {
    constructor(codeHash, args, hashType) {
        this.codeHash = codeHash;
        this.args = args;
        this.hashType = hashType;
    }
    static fromRPC(data) {
        if (!data)
            return undefined;
        validators.ValidateScript(data);
        return new Script(data.code_hash, data.args, data.hash_type);
    }
    sameWith(script) {
        validators.ValidateScript(transformers.TransformScript(script));
        return (this.args === script.args &&
            this.codeHash === script.codeHash &&
            this.hashType === script.hashType);
    }
    validate() {
        validators.ValidateScript(transformers.TransformScript(this));
        return true;
    }
    serializeJson() {
        return {
            code_hash: this.codeHash,
            args: this.args,
            hash_type: this.hashType,
        };
    }
    toHash() {
        return new Blake2bHasher()
            .update(SerializeScript(normalizers.NormalizeScript(transformers.TransformScript(this))))
            .digest()
            .serializeJson();
    }
    toAddress(prefix = getDefaultPrefix(), addressVersion = NervosAddressVersion.latest) {
        return new Address(generateCkbAddressString(this, prefix, addressVersion), AddressType.ckb);
    }
}
