// @lay2/pw-core/build/module/models/transaction.js
import { ECDSA_WITNESS_LEN } from '../constants';
import { validators, normalizers, transformers, Reader, } from '../ckb-js-toolkit';
import { SerializeTransaction, SerializeWitnessArgs } from '../ckb-lumos/core';
export class Transaction {
    constructor(raw, witnessArgs, _witnessLengths = [ECDSA_WITNESS_LEN]) {
        this.raw = raw;
        this.witnessArgs = witnessArgs;
        this.witnesses = raw.inputs.map((_) => '0x');
        // for (let i = 0; i < witnessLengths.length; i++) {
        //   this.witnesses[i] = '0x' + '0'.repeat(witnessLengths[i] - 2);
        // }
        if (!Array.isArray(witnessArgs))
            throw new Error('[Transaction] - witnessArgs must be an Array!');
        for (let i = 0; i < witnessArgs.length; i++) {
            if (typeof witnessArgs[i] !== 'string') {
                this.witnesses[i] = new Reader(SerializeWitnessArgs(normalizers.NormalizeWitnessArgs(this.witnessArgs[i]))).serializeJson();
            }
        }
    }
    sameWith(tx) {
        validators.ValidateTransaction(transformers.TransformTransaction(tx));
        return (tx.raw.sameWith(this.raw) &&
            tx.witnesses.join('-') === this.witnesses.join('-'));
    }
    getSize() {
        const tx = transformers.TransformTransaction(this);
        validators.ValidateTransaction(tx);
        // TODO: find out why the size is always smaller than the correct value by exact '4'
        return (SerializeTransaction(normalizers.NormalizeTransaction(tx)).byteLength + 4);
    }
    validate() {
        validators.ValidateTransaction(transformers.TransformTransaction(this));
        return this;
    }
    transform() {
        return transformers.TransformTransaction(this.serializeJson());
    }
    serializeJson() {
        return {
            ...this.raw,
            witnesses: this.witnesses,
        };
    }
}
