// @lay2/pw-core/build/main/models/amount.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Amount = exports.AmountUnit = void 0;
const jsbi_1 = __importDefault(require("jsbi"));
const utils_1 = require("../utils");
var AmountUnit;
(function (AmountUnit) {
    AmountUnit[AmountUnit["shannon"] = 0] = "shannon";
    AmountUnit[AmountUnit["ckb"] = 8] = "ckb";
})(AmountUnit = exports.AmountUnit || (exports.AmountUnit = {}));
class Amount {
    constructor(amount, decimals = AmountUnit.ckb) {
        if (!Number.isInteger(decimals) || decimals < 0) {
            throw new Error(`decimals ${decimals} must be a natural number`);
        }
        if (Number.isNaN(amount)) {
            throw new Error(`amount ${amount} must be a valid number`);
        }
        this.amount = utils_1.rationalNumberToBnString(amount, decimals);
    }
    add(val) {
        return new Amount(jsbi_1.default.add(this.toBigInt(), val.toBigInt()).toString(), 0);
    }
    sub(val) {
        return new Amount(jsbi_1.default.subtract(this.toBigInt(), val.toBigInt()).toString(), 0);
    }
    gt(val) {
        return jsbi_1.default.GT(this.toBigInt(), val.toBigInt());
    }
    gte(val) {
        return jsbi_1.default.greaterThanOrEqual(this.toBigInt(), val.toBigInt());
    }
    lt(val) {
        return jsbi_1.default.LT(this.toBigInt(), val.toBigInt());
    }
    lte(val) {
        return jsbi_1.default.lessThanOrEqual(this.toBigInt(), val.toBigInt());
    }
    eq(val) {
        return jsbi_1.default.EQ(this.toBigInt(), val.toBigInt());
    }
    toString(decimals = AmountUnit.ckb, options) {
        return utils_1.bnStringToRationalNumber(this.toBigInt().toString(), decimals, options);
    }
    toBigInt() {
        return jsbi_1.default.BigInt(this.amount);
    }
    toHexString() {
        return `0x${this.toBigInt().toString(16)}`;
    }
    toUInt128LE() {
        return utils_1.toBigUInt128LE(jsbi_1.default.BigInt(this.toHexString()));
    }
    static fromUInt128LE(hex) {
        return new Amount(`0x${utils_1.readBigUInt128LE(hex).toString(16)}`, AmountUnit.shannon);
    }
}
exports.Amount = Amount;
Amount.ZERO = new Amount('0');
