// @lay2/pw-core/build/main/models/cell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cell = void 0;
const interfaces_1 = require("../interfaces");
const script_1 = require("./script");
const out_point_1 = require("./out-point");
const cell_input_1 = require("./cell-input");
// import { minimalCellCapacity } from '../utils';
const amount_1 = require("./amount");
const ckb_js_toolkit_1 = require("../ckb-js-toolkit");
const utils_1 = require("../utils");
class Cell {
    constructor(capacity, lock, type, outPoint, data = '0x') {
        this.capacity = capacity;
        this.lock = lock;
        this.type = type;
        this.outPoint = outPoint;
        this.data = data;
        this.spaceCheck();
    }
    static fromRPC(data) {
        if (!data) {
            throw new Error('Cannot create cell from empty data');
        }
        ckb_js_toolkit_1.validators.ValidateCellOutput(data);
        return new Cell(data.capacity, script_1.Script.fromRPC(data.lock), script_1.Script.fromRPC(data.type), out_point_1.OutPoint.fromRPC(data.out_point), data.data);
    }
    static async loadFromBlockchain(rpc, outPoint) {
        const index = Number(outPoint.index);
        const { transaction: { outputs, outputs_data }, } = await rpc.get_transaction(outPoint.txHash);
        const { capacity, lock, type } = outputs[index];
        return new Cell(new amount_1.Amount(capacity, amount_1.AmountUnit.shannon), new script_1.Script(lock.code_hash, lock.args, interfaces_1.HashType[lock.hash_type]), type
            ? new script_1.Script(type.code_hash, type.args, interfaces_1.HashType[type.hash_type])
            : null, outPoint, outputs_data[index]);
    }
    clone() {
        return new Cell(this.capacity, this.lock, this.type, this.outPoint, this.data);
    }
    sameWith(cell) {
        if (!cell || !cell.outPoint || !this.outPoint) {
            throw new Error('to be compared, cells must have outPoint value');
        }
        return cell.outPoint.sameWith(this.outPoint);
    }
    resize() {
        // const base = SerializeCellOutput(
        //   normalizers.NormalizeCellOutput(transformers.TransformCellOutput(this))
        // ).byteLength;
        const base = this.type ? 102 : 61;
        const extra = new Buffer(this.data.replace('0x', ''), 'hex').byteLength;
        const size = base + extra;
        this.capacity = new amount_1.Amount(size.toString());
        return size;
    }
    spaceCheck() {
        if (this.capacity.lt(this.occupiedCapacity())) {
            throw new Error(`cell capacity ${this.capacity.toString(amount_1.AmountUnit.ckb)} less than the min capacity ${this.occupiedCapacity().toString(amount_1.AmountUnit.ckb)}`);
        }
        return true;
    }
    occupiedCapacity() {
        return new amount_1.Amount(utils_1.cellOccupiedBytes(this).toString(), amount_1.AmountUnit.ckb);
    }
    availableFee() {
        return this.capacity.sub(this.occupiedCapacity());
    }
    toCellInput(since = '0x0') {
        return this.outPoint ? new cell_input_1.CellInput(this.outPoint, since) : undefined;
    }
    validate() {
        ckb_js_toolkit_1.validators.ValidateCellOutput(ckb_js_toolkit_1.transformers.TransformCellOutput(this));
        if (this.outPoint) {
            ckb_js_toolkit_1.validators.ValidateCellInput(ckb_js_toolkit_1.transformers.TransformCellInput(this.toCellInput()));
        }
        return this;
    }
    // CellOutput format
    serializeJson() {
        return {
            capacity: this.capacity.toHexString(),
            lock: this.lock,
            type: this.type,
        };
    }
    setData(data) {
        data = data.trim();
        const bytes = [];
        for (let i = 0; i < data.length; i++) {
            bytes.push(data.charCodeAt(i));
        }
        this.data = utils_1.byteArrayToHex(bytes);
        this.spaceCheck();
    }
    setHexData(data) {
        data = data.trim();
        if (!data.startsWith('0x')) {
            throw new Error('Hex data should start with 0x');
        }
        this.data = data;
        this.spaceCheck();
    }
    getData() {
        return utils_1.hexToByteArray(this.data.trim())
            .map((char) => String.fromCharCode(char))
            .join('');
    }
    getHexData() {
        return this.data.trim();
    }
    setSUDTAmount(amount) {
        this.data = amount.toUInt128LE() + this.data.slice(34);
    }
    getSUDTAmount() {
        const sudtAmountData = this.data.slice(0, 34);
        return amount_1.Amount.fromUInt128LE(sudtAmountData);
    }
    isEmpty() {
        return this.data.trim() === '0x';
    }
}
exports.Cell = Cell;
