// @lay2/pw-core/build/main/models/cell-input.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CellInput = void 0;
const ckb_js_toolkit_1 = require("../ckb-js-toolkit");
class CellInput {
    constructor(previousOutput, since = '0x0') {
        this.previousOutput = previousOutput;
        this.since = since;
    }
    static fromRPC(data) {
        if (!data) {
            throw new Error('Cannot create CellInput from empty data');
        }
        ckb_js_toolkit_1.validators.ValidateCellInput(data);
        return new CellInput(data.previous_output, data.since);
    }
    sameWith(cellInput) {
        ckb_js_toolkit_1.validators.ValidateCellInput(ckb_js_toolkit_1.transformers.TransformCellInput(cellInput));
        return (cellInput.previousOutput.sameWith(this.previousOutput) &&
            cellInput.since === this.since);
    }
    validate() {
        ckb_js_toolkit_1.validators.ValidateCellInput(ckb_js_toolkit_1.transformers.TransformCellInput(this));
        return true;
    }
    serializeJson() {
        return {
            since: this.since,
            previous_output: this.previousOutput,
        };
    }
}
exports.CellInput = CellInput;
