// @lay2/pw-core/build/main/models/cell-dep.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CellDep = void 0;
const ckb_js_toolkit_1 = require("../ckb-js-toolkit");
class CellDep {
    constructor(depType, outPoint) {
        this.depType = depType;
        this.outPoint = outPoint;
    }
    static fromRPC(data) {
        if (!data)
            return undefined;
        ckb_js_toolkit_1.validators.ValidateCellDep(data);
        return new CellDep(data.dep_type, data.out_point);
    }
    validate() {
        ckb_js_toolkit_1.validators.ValidateCellDep(ckb_js_toolkit_1.transformers.TransformCellDep(this));
        return true;
    }
    sameWith(cellDep) {
        ckb_js_toolkit_1.validators.ValidateCellDep(ckb_js_toolkit_1.transformers.TransformCellDep(cellDep));
        return (cellDep.depType === this.depType &&
            cellDep.outPoint.sameWith(this.outPoint));
    }
    serializeJson() {
        return {
            dep_type: this.depType,
            out_point: this.outPoint,
        };
    }
}
exports.CellDep = CellDep;
