// @lay2/pw-core/build/main/models/out-point.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OutPoint = void 0;
const ckb_js_toolkit_1 = require("../ckb-js-toolkit");
class OutPoint {
    constructor(txHash, index) {
        this.txHash = txHash;
        this.index = index;
    }
    static fromRPC(data) {
        if (!data)
            return undefined;
        ckb_js_toolkit_1.validators.ValidateOutPoint(data);
        return new OutPoint(data.tx_hash, data.index);
    }
    sameWith(outPoint) {
        ckb_js_toolkit_1.validators.ValidateOutPoint(ckb_js_toolkit_1.transformers.TransformOutPoint(outPoint));
        return this.txHash === outPoint.txHash && this.index === outPoint.index;
    }
    validate() {
        ckb_js_toolkit_1.validators.ValidateOutPoint(ckb_js_toolkit_1.transformers.TransformOutPoint(this));
        return this;
    }
    serializeJson() {
        return {
            tx_hash: this.txHash,
            index: this.index,
        };
    }
}
exports.OutPoint = OutPoint;
