// @lay2/pw-core/build/main/models/raw-transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RawTransaction = void 0;
const __1 = __importDefault(require(".."));
const core_1 = require("../ckb-lumos/core");
const ckb_js_toolkit_1 = require("../ckb-js-toolkit");
const hashers_1 = require("../hashers");
class RawTransaction {
    constructor(inputCells, outputs, cellDeps = [
        __1.default.config.defaultLock.cellDep,
        __1.default.config.pwLock.cellDep,
    ], headerDeps = [], version = '0x0') {
        this.inputCells = inputCells;
        this.outputs = outputs;
        this.cellDeps = cellDeps;
        this.headerDeps = headerDeps;
        this.version = version;
        this.inputs = inputCells.map((i) => i.toCellInput());
        this.outputsData = this.outputs.map((o) => o.getHexData());
    }
    sameWith(raw) {
        ckb_js_toolkit_1.validators.ValidateTransaction(ckb_js_toolkit_1.transformers.TransformTransaction(raw));
        return raw.toHash() === this.toHash();
    }
    toHash() {
        const rawTx = ckb_js_toolkit_1.transformers.TransformRawTransaction(this);
        const hasher = new hashers_1.Blake2bHasher();
        return hasher
            .hash(new ckb_js_toolkit_1.Reader(core_1.SerializeRawTransaction(ckb_js_toolkit_1.normalizers.NormalizeRawTransaction(rawTx))))
            .serializeJson();
    }
    validate() {
        ckb_js_toolkit_1.validators.ValidateRawTransaction(ckb_js_toolkit_1.transformers.TransformRawTransaction(this));
        return true;
    }
    serializeJson() {
        return {
            version: this.version,
            cell_deps: this.cellDeps,
            header_deps: this.headerDeps,
            inputs: this.inputs,
            outputs: this.outputs,
            outputs_data: this.outputsData,
        };
    }
}
exports.RawTransaction = RawTransaction;
