// @lay2/pw-core/build/main/models/script.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Script = void 0;
const address_1 = require("./address");
const utils_1 = require("../utils");
const ckb_js_toolkit_1 = require("../ckb-js-toolkit");
const core_1 = require("../ckb-lumos/core");
const hashers_1 = require("../hashers");
const address_2 = require("../helpers/address");
class Script {
    constructor(codeHash, args, hashType) {
        this.codeHash = codeHash;
        this.args = args;
        this.hashType = hashType;
    }
    static fromRPC(data) {
        if (!data)
            return undefined;
        ckb_js_toolkit_1.validators.ValidateScript(data);
        return new Script(data.code_hash, data.args, data.hash_type);
    }
    sameWith(script) {
        ckb_js_toolkit_1.validators.ValidateScript(ckb_js_toolkit_1.transformers.TransformScript(script));
        return (this.args === script.args &&
            this.codeHash === script.codeHash &&
            this.hashType === script.hashType);
    }
    validate() {
        ckb_js_toolkit_1.validators.ValidateScript(ckb_js_toolkit_1.transformers.TransformScript(this));
        return true;
    }
    serializeJson() {
        return {
            code_hash: this.codeHash,
            args: this.args,
            hash_type: this.hashType,
        };
    }
    toHash() {
        return new hashers_1.Blake2bHasher()
            .update(core_1.SerializeScript(ckb_js_toolkit_1.normalizers.NormalizeScript(ckb_js_toolkit_1.transformers.TransformScript(this))))
            .digest()
            .serializeJson();
    }
    toAddress(prefix = address_1.getDefaultPrefix(), addressVersion = address_2.NervosAddressVersion.latest) {
        return new address_1.Address(utils_1.generateCkbAddressString(this, prefix, addressVersion), address_1.AddressType.ckb);
    }
}
exports.Script = Script;
