// @lay2/pw-core/build/main/models/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const constants_1 = require("../constants");
const ckb_js_toolkit_1 = require("../ckb-js-toolkit");
const core_1 = require("../ckb-lumos/core");
class Transaction {
    constructor(raw, witnessArgs, _witnessLengths = [constants_1.ECDSA_WITNESS_LEN]) {
        this.raw = raw;
        this.witnessArgs = witnessArgs;
        this.witnesses = raw.inputs.map((_) => '0x');
        // for (let i = 0; i < witnessLengths.length; i++) {
        //   this.witnesses[i] = '0x' + '0'.repeat(witnessLengths[i] - 2);
        // }
        if (!Array.isArray(witnessArgs))
            throw new Error('[Transaction] - witnessArgs must be an Array!');
        for (let i = 0; i < witnessArgs.length; i++) {
            if (typeof witnessArgs[i] !== 'string') {
                this.witnesses[i] = new ckb_js_toolkit_1.Reader(core_1.SerializeWitnessArgs(ckb_js_toolkit_1.normalizers.NormalizeWitnessArgs(this.witnessArgs[i]))).serializeJson();
            }
        }
    }
    sameWith(tx) {
        ckb_js_toolkit_1.validators.ValidateTransaction(ckb_js_toolkit_1.transformers.TransformTransaction(tx));
        return (tx.raw.sameWith(this.raw) &&
            tx.witnesses.join('-') === this.witnesses.join('-'));
    }
    getSize() {
        const tx = ckb_js_toolkit_1.transformers.TransformTransaction(this);
        ckb_js_toolkit_1.validators.ValidateTransaction(tx);
        // TODO: find out why the size is always smaller than the correct value by exact '4'
        return (core_1.SerializeTransaction(ckb_js_toolkit_1.normalizers.NormalizeTransaction(tx)).byteLength + 4);
    }
    validate() {
        ckb_js_toolkit_1.validators.ValidateTransaction(ckb_js_toolkit_1.transformers.TransformTransaction(this));
        return this;
    }
    transform() {
        return ckb_js_toolkit_1.transformers.TransformTransaction(this.serializeJson());
    }
    serializeJson() {
        return Object.assign(Object.assign({}, this.raw), { witnesses: this.witnesses });
    }
}
exports.Transaction = Transaction;
