// @lay2/pw-core/build/module/helpers/ckb-indexer.js
import axios from 'axios';
import { Cell, Amount, AmountUnit, OutPoint, Script as PwScript, } from '..';
export var ScriptType;
(function (ScriptType) {
    ScriptType["type"] = "type";
    ScriptType["lock"] = "lock";
})(ScriptType || (ScriptType = {}));
export var Order;
(function (Order) {
    Order["asc"] = "asc";
    Order["desc"] = "desc";
})(Order || (Order = {}));
export const DefaultTerminator = (_index, _cell) => {
    return { stop: false, push: true };
};
export class CkbIndexer {
    constructor(ckbIndexerUrl) {
        this.ckbIndexerUrl = ckbIndexerUrl;
    }
    async request(method, params) {
        const data = {
            id: 0,
            jsonrpc: '2.0',
            method,
            params,
        };
        const res = await axios.post(this.ckbIndexerUrl, data);
        if (res.status !== 200) {
            throw new Error(`indexer request failed with HTTP code ${res.status}`);
        }
        if (res.data.error !== undefined) {
            throw new Error(`indexer request rpc failed with error: ${JSON.stringify(res.data.error)}`);
        }
        return res.data.result;
    }
    async getCells(searchKey, terminator = DefaultTerminator, { sizeLimit = 0x100, order = Order.asc, } = {}) {
        const infos = [];
        let cursor = null;
        let index = 0;
        const params = [searchKey, order, `0x${sizeLimit.toString(16)}`, cursor];
        while (true) {
            const res = await this.request('get_cells', params);
            const liveCells = res.objects;
            cursor = res.lastCursor;
            for (const cell of liveCells) {
                const { stop, push } = terminator(index, cell);
                if (push) {
                    infos.push(cell);
                }
                if (stop) {
                    return infos;
                }
                index++;
            }
            if (liveCells.length < sizeLimit) {
                break;
            }
        }
        return infos;
    }
}
export function IndexerCellToCell(cell) {
    return new Cell(new Amount(cell.output.capacity, AmountUnit.shannon), PwScript.fromRPC(cell.output.lock), PwScript.fromRPC(cell.output.type), OutPoint.fromRPC(cell.out_point), cell.output_data);
}
