// @ckb-lumos/ckb-indexer/lib/rpc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RPC = void 0;

var _toolkit = require("@ckb-lumos/toolkit");

const handler = {
  get: (target, method) => {
    return async (...params) => {
      const result = await target.rpc[method](...params);
      return result;
    };
  }
};

class RpcProxy {
  constructor(uri) {
    this.rpc = new _toolkit.RPC(uri);
  }

  getProxy() {
    return new Proxy(this, handler);
  }

}

class RPC {
  /**
   *
   * @param uri  indexer uri
   */
  constructor(uri) {
    this.rpcProxy = new RpcProxy(uri).getProxy();
  }

  async get_tip() {
    return this.rpcProxy.get_tip();
  }

  async get_cells(searchKey, order, limit, after_cursor) {
    return this.rpcProxy.get_cells(searchKey, order, limit, after_cursor);
  }

  async get_transactions(searchKey, order, limit, after_cursor) {
    return this.rpcProxy.get_transactions(searchKey, order, limit, after_cursor);
  }

  async get_cells_capacity(searchKey) {
    return this.rpcProxy.get_cells_capacity(searchKey);
  }

  async get_indexer_info() {
    return this.rpcProxy.get_indexer_info();
  }

}

exports.RPC = RPC;
