// @lay2/pw-core/build/main/builders/builder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Builder = void 0;
const models_1 = require("../models");
const __1 = __importStar(require(".."));
const FEE_BASE = 1000;
class Builder {
    constructor(feeRate = Builder.MIN_FEE_RATE, collector = __1.default.defaultCollector, witnessArgs = __1.default.chainId === __1.ChainID.ckb
        ? Builder.WITNESS_ARGS.RawSecp256k1
        : Builder.WITNESS_ARGS.Secp256k1) {
        this.feeRate = feeRate;
        this.collector = collector;
        this.witnessArgs = witnessArgs;
    }
    static calcFee(tx, feeRate = Builder.MIN_FEE_RATE) {
        if (feeRate < Builder.MIN_FEE_RATE) {
            feeRate = Builder.MIN_FEE_RATE;
        }
        const txSize = tx.getSize();
        const fee = (feeRate / FEE_BASE) * txSize;
        return new models_1.Amount(fee.toString(), models_1.AmountUnit.shannon);
    }
    getFee() {
        return this.fee;
    }
}
exports.Builder = Builder;
Builder.MIN_FEE_RATE = 1000;
Builder.MIN_CHANGE = new models_1.Amount('61', models_1.AmountUnit.ckb);
Builder.WITNESS_ARGS = {
    Secp256k1: {
        lock: '0x' + '0'.repeat(132),
        input_type: '',
        output_type: '',
    },
    RawSecp256k1: {
        lock: '0x' + '0'.repeat(130),
        input_type: '',
        output_type: '',
    },
    Secp256r1: {
        lock: '0x' + '0'.repeat(600),
        input_type: '',
        output_type: '',
    },
};
