// @lay2/pw-core/build/main/builders/simple-builder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleBuilder = void 0;
const builder_1 = require("../builders/builder");
const models_1 = require("../models");
const core_1 = __importDefault(require("../core"));
const simple_acp_builder_1 = require("./simple-acp-builder");
class SimpleBuilder extends builder_1.Builder {
    constructor(address, amount, options = {}) {
        super(options.feeRate, options.collector, options.witnessArgs);
        this.address = address;
        this.amount = amount;
        this.options = options;
        this.simpleACPBuilder = new simple_acp_builder_1.SimpleACPBuilder(this.address, this.amount, this.options);
    }
    async build(fee = models_1.Amount.ZERO) {
        if (this.amount.lt(builder_1.Builder.MIN_CHANGE)) {
            return this.simpleACPBuilder.build();
        }
        const outputCell = new models_1.Cell(this.amount, this.address.toLockScript());
        const data = this.options.data;
        if (data) {
            if (data.startsWith('0x')) {
                outputCell.setHexData(data);
            }
            else {
                outputCell.setData(data);
            }
        }
        const neededAmount = this.amount.add(builder_1.Builder.MIN_CHANGE).add(fee);
        let inputSum = new models_1.Amount('0');
        const inputCells = [];
        // fill the inputs
        const cells = await this.collector.collect(core_1.default.provider.address, {
            neededAmount,
        });
        for (const cell of cells) {
            inputCells.push(cell);
            inputSum = inputSum.add(cell.capacity);
            if (inputSum.gt(neededAmount))
                break;
        }
        if (inputSum.lt(neededAmount)) {
            throw new Error(`input capacity not enough, need ${neededAmount.toString(models_1.AmountUnit.ckb)}, got ${inputSum.toString(models_1.AmountUnit.ckb)}`);
        }
        const changeCell = new models_1.Cell(inputSum.sub(outputCell.capacity), core_1.default.provider.address.toLockScript());
        const tx = new models_1.Transaction(new models_1.RawTransaction(inputCells, [outputCell, changeCell]), [this.witnessArgs]);
        this.fee = builder_1.Builder.calcFee(tx, this.feeRate);
        if (changeCell.capacity.gte(builder_1.Builder.MIN_CHANGE.add(this.fee))) {
            changeCell.capacity = changeCell.capacity.sub(this.fee);
            tx.raw.outputs.pop();
            tx.raw.outputs.push(changeCell);
            return tx;
        }
        return this.build(this.fee);
    }
    getCollector() {
        return this.collector;
    }
}
exports.SimpleBuilder = SimpleBuilder;
