// @lay2/pw-core/build/main/builders/simple-acp-builder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleACPBuilder = void 0;
const builder_1 = require("../builders/builder");
const models_1 = require("../models");
const __1 = __importDefault(require(".."));
class SimpleACPBuilder extends builder_1.Builder {
    constructor(address, amount, options = {}) {
        super(options.feeRate, options.collector, options.witnessArgs);
        this.address = address;
        this.amount = amount;
        this.options = options;
    }
    async build() {
        if (!this.address.isAcp()) {
            throw new Error("The Receiver's address is not anyone-can-pay cell");
        }
        const receiverACPCells = await this.collector.collect(this.address, {
            neededAmount: new models_1.Amount('1', models_1.AmountUnit.shannon),
        });
        if (!receiverACPCells || receiverACPCells.length === 0) {
            throw new Error('The receiver has no sudt cell');
        }
        this.receiverInputCell = receiverACPCells[0];
        this.receiverOutputCell = this.receiverInputCell.clone();
        this.receiverOutputCell.capacity = this.receiverOutputCell.capacity.add(this.amount);
        return this.buildSenderCells();
    }
    async buildSenderCells(fee = models_1.Amount.ZERO) {
        const neededAmount = this.amount.add(builder_1.Builder.MIN_CHANGE).add(fee);
        let inputSum = new models_1.Amount('0');
        const inputCells = [];
        // fill the inputs
        const cells = await this.collector.collect(__1.default.provider.address, {
            neededAmount,
        });
        for (const cell of cells) {
            inputCells.push(cell);
            inputSum = inputSum.add(cell.capacity);
            if (inputSum.gt(neededAmount))
                break;
        }
        if (inputSum.lt(neededAmount)) {
            throw new Error(`input capacity not enough, need ${neededAmount.toString(models_1.AmountUnit.ckb)}, got ${inputSum.toString(models_1.AmountUnit.ckb)}`);
        }
        const changeCell = new models_1.Cell(inputSum.sub(this.amount), __1.default.provider.address.toLockScript());
        const tx = new models_1.Transaction(new models_1.RawTransaction([...inputCells, this.receiverInputCell], [this.receiverOutputCell, changeCell]), [this.witnessArgs]);
        this.fee = builder_1.Builder.calcFee(tx, this.feeRate);
        if (changeCell.capacity.gte(builder_1.Builder.MIN_CHANGE.add(this.fee))) {
            changeCell.capacity = changeCell.capacity.sub(this.fee);
            tx.raw.outputs.pop();
            tx.raw.outputs.push(changeCell);
            return tx;
        }
        return this.buildSenderCells(this.fee);
    }
    getCollector() {
        return this.collector;
    }
}
exports.SimpleACPBuilder = SimpleACPBuilder;
