// @lay2/pw-core/build/main/builders/simple-sudt-acp-builder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleSUDTACPBuilder = void 0;
const builder_1 = require("./builder");
const models_1 = require("../models");
const __1 = __importDefault(require(".."));
const sudt_collector_1 = require("../collectors/sudt-collector");
class SimpleSUDTACPBuilder extends builder_1.Builder {
    constructor(sudt, address, amount, options = {}) {
        super(options.feeRate, options.collector, options.witnessArgs);
        this.sudt = sudt;
        this.address = address;
        this.amount = amount;
        this.options = options;
    }
    async build() {
        if (!this.address.isAcp()) {
            throw new Error("The receiver's address is not anyone-can-pay cell");
        }
        if (!(this.collector instanceof sudt_collector_1.SUDTCollector)) {
            throw new Error('this.collector is not a SUDTCollector instance');
        }
        const receiverSUDTCells = await this.collector.collectSUDT(this.sudt, this.address, { neededAmount: new models_1.Amount('1', models_1.AmountUnit.shannon) });
        if (!receiverSUDTCells || receiverSUDTCells.length === 0) {
            throw new Error('The receiver has no sudt cell');
        }
        const receiverInputCell = receiverSUDTCells[0];
        const receiverOuputCell = receiverInputCell.clone();
        receiverOuputCell.setSUDTAmount(this.amount.add(receiverOuputCell.getSUDTAmount()));
        let senderInputSUDTSum = new models_1.Amount('0');
        let senderInputCKBSum = new models_1.Amount('0');
        let minSenderOccupiedCKBSum = new models_1.Amount('0');
        let restNeededSUDT = new models_1.Amount(this.amount.toHexString(), models_1.AmountUnit.shannon);
        const inputCells = [];
        const outputCells = [];
        // fill the inputs and the outputs
        const unspentSUDTCells = await this.collector.collectSUDT(this.sudt, __1.default.provider.address, { neededAmount: this.amount });
        // First step: build a tx including sender and receiver sudt cell only
        for (const inputCell of unspentSUDTCells) {
            const outputCell = inputCell.clone();
            const inputSUDTAmount = inputCell.getSUDTAmount();
            senderInputSUDTSum = senderInputSUDTSum.add(inputSUDTAmount);
            senderInputCKBSum = senderInputCKBSum.add(inputCell.capacity);
            minSenderOccupiedCKBSum = minSenderOccupiedCKBSum.add(outputCell.occupiedCapacity());
            if (inputSUDTAmount.lt(restNeededSUDT)) {
                restNeededSUDT = restNeededSUDT.sub(inputSUDTAmount);
                outputCell.setSUDTAmount(new models_1.Amount('0'));
            }
            else {
                outputCell.setSUDTAmount(inputSUDTAmount.sub(restNeededSUDT));
                restNeededSUDT = new models_1.Amount('0');
            }
            inputCells.push(inputCell);
            outputCells.unshift(outputCell);
            if (senderInputSUDTSum.gte(this.amount))
                break;
        }
        if (senderInputSUDTSum.lt(this.amount)) {
            throw new Error(`input sudt amount not enough, need ${this.amount.toString(models_1.AmountUnit.ckb)}, got ${senderInputSUDTSum.toString(models_1.AmountUnit.ckb)}`);
        }
        inputCells.push(receiverInputCell);
        outputCells.unshift(receiverOuputCell);
        let tx = this.rectifyTx(inputCells, outputCells);
        const availableCKBFee = senderInputCKBSum.sub(minSenderOccupiedCKBSum);
        // Second step:  if sudt cell can not pay the transaction fee, fetch pure ckb cells to pay the fee.
        if (this.fee.gt(availableCKBFee)) {
            const unspentCKBCells = await this.collector.collect(__1.default.provider.address, { neededAmount: this.fee.sub(availableCKBFee).add(builder_1.Builder.MIN_CHANGE) });
            if (!unspentCKBCells || unspentCKBCells.length === 0) {
                throw new Error('not enough CKB to pay the transaction fee');
            }
            // append the fee cell to tx's inputs and outputs
            const ckbFeeInputCell = unspentCKBCells[0];
            inputCells.push(ckbFeeInputCell);
            outputCells.push(ckbFeeInputCell.clone());
            tx = this.rectifyTx(inputCells, outputCells);
            // if fee change cell's capacity less than occuiped capacity, merge the fee cell to sender's input sudt cell.
            if (this.fee.gt(availableCKBFee.add(ckbFeeInputCell.availableFee()))) {
                outputCells.pop();
                const senderOutputCell = outputCells.pop();
                senderOutputCell.capacity = senderOutputCell.capacity.add(ckbFeeInputCell.capacity);
                outputCells.push(senderOutputCell);
                tx = this.rectifyTx(inputCells, outputCells);
            }
        }
        // Third step: subtract tx fee from outputs' capacity
        tx = this.subtractFee(inputCells, outputCells);
        return tx;
    }
    subtractFee(inputCells, outputCells) {
        let remainFee = new models_1.Amount(this.fee.toHexString(), models_1.AmountUnit.shannon);
        for (const cell of outputCells.slice(1)) {
            // throw new Error(`remainFee ${remainFee} ${cell.availableFee()}`);
            if (remainFee.gt(cell.availableFee())) {
                remainFee = remainFee.sub(cell.availableFee());
                cell.capacity = cell.occupiedCapacity();
            }
            else {
                cell.capacity = cell.capacity.sub(remainFee);
                break;
            }
        }
        return this.rectifyTx(inputCells, outputCells);
    }
    rectifyTx(inputCells, outputCells) {
        const sudtCellDeps = [
            __1.default.config.defaultLock.cellDep,
            __1.default.config.pwLock.cellDep,
            __1.default.config.sudtType.cellDep,
        ];
        const tx = new models_1.Transaction(new models_1.RawTransaction(inputCells, outputCells, sudtCellDeps), [this.witnessArgs]);
        this.fee = builder_1.Builder.calcFee(tx, this.feeRate);
        return tx;
    }
    getCollector() {
        return this.collector;
    }
}
exports.SimpleSUDTACPBuilder = SimpleSUDTACPBuilder;
