// @lay2/pw-core/build/module/core.js
import { RPC, transformers } from './ckb-js-toolkit';
import { CHAIN_SPECS } from './constants';
import { DefaultSigner } from './signers';
import { Collector } from './collectors';
import { SimpleBuilder, Builder, SimpleSUDTACPBuilder, SimpleSUDTBuilder, } from './builders';
import { Provider } from './providers';
export var ChainID;
(function (ChainID) {
    ChainID[ChainID["ckb"] = 0] = "ckb";
    ChainID[ChainID["ckb_testnet"] = 1] = "ckb_testnet";
    ChainID[ChainID["ckb_dev"] = 2] = "ckb_dev";
})(ChainID || (ChainID = {}));
/**
 * The default main class of pw-core
 */
export default class PWCore {
    constructor(nodeUrl) {
        this._rpc = new RPC(nodeUrl);
    }
    /**
     * set chain id for pw-core
     *
     * @param chainId chain id of ckb blockchain: 0 for mainnet, 1 for testnet, 2 for devnet
     * @param config chain config
     */
    static setChainId(chainId, config) {
        if (chainId !== ChainID.ckb && chainId !== ChainID.ckb_testnet && !config) {
            throw new Error('config must be provided for dev chain');
        }
        PWCore.chainId = chainId;
        PWCore.config = config ? config : CHAIN_SPECS[chainId];
    }
    /**
     * Initialize the environment required by pw-core
     */
    async init(provider, defaultCollector, chainId, config) {
        if (chainId !== undefined) {
            if (!(chainId in ChainID)) {
                throw new Error(`invalid chainId ${chainId}`);
            }
            PWCore.chainId = chainId;
        }
        else {
            const info = await this.rpc.get_blockchain_info();
            PWCore.chainId = {
                ckb: ChainID.ckb,
                ckb_testnet: ChainID.ckb_testnet,
                ckb_dev: ChainID.ckb_dev,
            }[info.chain];
        }
        if (PWCore.chainId === ChainID.ckb_dev) {
            if (!config) {
                throw new Error('config must be provided for dev chain');
            }
            PWCore.config = config;
        }
        else {
            // merge customized config to default one
            PWCore.config = {
                ...[CHAIN_SPECS.Lina, CHAIN_SPECS.Aggron][PWCore.chainId],
                ...config,
            };
        }
        if (provider instanceof Provider) {
            PWCore.provider = await provider.init();
        }
        else {
            throw new Error('provider must be provided');
        }
        if (defaultCollector instanceof Collector) {
            PWCore.defaultCollector = defaultCollector;
        }
        else {
            throw new Error('defaultCollector must be provided');
        }
        return this;
    }
    /**
     * Return a RPC instance defined in package 'ckb-js-toolkit'
     */
    get rpc() {
        return this._rpc;
    }
    /**
     * Transfer CKB to any address
     * @param address The receiver's address
     * @param amount The amount of CKB to send
     * @param options The transaction builder options for this transaction.
     */
    async send(address, amount, options) {
        const simpleBuilder = new SimpleBuilder(address, amount, options);
        return this.sendTransaction(simpleBuilder);
    }
    /**
     * Send an built transaction or a builder
     * @param toSend
     * @param signer
     */
    async sendTransaction(toSend, signer) {
        const tx = toSend instanceof Builder ? await toSend.build() : toSend;
        tx.validate();
        if (!signer) {
            signer = new DefaultSigner(PWCore.provider);
        }
        return this.rpc.send_transaction(transformers.TransformTransaction(await signer.sign(tx)), 'passthrough');
    }
    /**
     * Transfer sudt to any address
     * @param sudt The sudt definition
     * @param address the receiver's address
     * @param amount the aount of sudt to send
     * @param options The transaction builder options for this transaction.
     * @returns the transaction hash
     */
    async sendSUDT(sudt, address, amount, createAcp, signer, options) {
        const builder = createAcp
            ? new SimpleSUDTBuilder(sudt, address, amount, options)
            : new SimpleSUDTACPBuilder(sudt, address, amount, options);
        return this.sendTransaction(builder, signer);
    }
}
PWCore.config = CHAIN_SPECS[ChainID.ckb_testnet];
PWCore.chainId = ChainID.ckb_testnet;
