// @lay2/pw-core/build/main/collectors/pw-collector.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PwCollector = void 0;
const axios_1 = __importDefault(require("axios"));
const sudt_collector_1 = require("./sudt-collector");
const __1 = require("..");
class PwCollector extends sudt_collector_1.SUDTCollector {
    constructor(apiBase) {
        super();
        this.apiBase = apiBase;
        this.apiBase = apiBase;
    }
    async getBalance(address) {
        const res = await axios_1.default.get(`${this.apiBase}/cell/getCapacityByLockHash?lockHash=${address
            .toLockScript()
            .toHash()}`);
        return new __1.Amount(res.data.data, __1.AmountUnit.shannon);
    }
    async collect(address, options) {
        if (!options || !options.neededAmount) {
            throw new Error("'neededAmount' in options must be provided");
        }
        const cells = [];
        const res = await axios_1.default.get(`${this.apiBase}/cell/unSpent?lockHash=${address
            .toLockScript()
            .toHash()}&capacity=${options.neededAmount.toHexString()}`);
        for (let { capacity, outPoint } of res.data.data) {
            capacity = new __1.Amount(capacity, __1.AmountUnit.shannon);
            outPoint = new __1.OutPoint(outPoint.txHash, outPoint.index);
            cells.push(new __1.Cell(capacity, address.toLockScript(), null, outPoint));
        }
        return cells;
    }
    async getSUDTBalance(sudt, address) {
        const lockHash = address.toLockScript().toHash();
        const typeHash = sudt.toTypeScript().toHash();
        const res = await axios_1.default.get(`${this.apiBase}/sudt/balance?lockHash=${lockHash}&typeHash=${typeHash}`);
        return new __1.Amount(res.data.data.sudtAmount, __1.AmountUnit.shannon);
    }
    async collectSUDT(sudt, address, options) {
        if (!options || !options.neededAmount) {
            throw new Error("'neededAmount' in options must be provided");
        }
        const cells = [];
        const lockHash = address.toLockScript().toHash();
        const typeHash = sudt.toTypeScript().toHash();
        const res = await axios_1.default.get(`${this.apiBase}/cell/unSpent?lockHash=${lockHash}&capacity=0x0&typeHash=${typeHash}&sudtAmount=${options.neededAmount.toHexString()}`);
        for (let { capacity, outPoint, type, data } of res.data.data) {
            capacity = new __1.Amount(capacity, __1.AmountUnit.shannon);
            outPoint = new __1.OutPoint(outPoint.txHash, outPoint.index);
            cells.push(new __1.Cell(capacity, address.toLockScript(), type, outPoint, data));
        }
        return cells;
    }
}
exports.PwCollector = PwCollector;
