// @polyjuice-provider/base/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FailedReasonStatusType = exports.SigningMessageType = exports.ShortAddressType = void 0;
var ShortAddressType;
(function (ShortAddressType) {
    ShortAddressType[ShortAddressType["eoaAddress"] = 0] = "eoaAddress";
    ShortAddressType[ShortAddressType["contractAddress"] = 1] = "contractAddress";
    ShortAddressType[ShortAddressType["notExistEoaAddress"] = 2] = "notExistEoaAddress";
    ShortAddressType[ShortAddressType["notExistContractAddress"] = 3] = "notExistContractAddress";
    ShortAddressType[ShortAddressType["creatorAddress"] = 4] = "creatorAddress";
})(ShortAddressType = exports.ShortAddressType || (exports.ShortAddressType = {}));
var SigningMessageType;
(function (SigningMessageType) {
    SigningMessageType[SigningMessageType["withPrefix"] = 0] = "withPrefix";
    SigningMessageType[SigningMessageType["noPrefix"] = 1] = "noPrefix";
})(SigningMessageType = exports.SigningMessageType || (exports.SigningMessageType = {}));
var FailedReasonStatusType;
(function (FailedReasonStatusType) {
    FailedReasonStatusType[FailedReasonStatusType["SUCCESS"] = 0] = "SUCCESS";
    FailedReasonStatusType[FailedReasonStatusType["FAILURE"] = 1] = "FAILURE";
    FailedReasonStatusType[FailedReasonStatusType["REVERT"] = 2] = "REVERT";
    FailedReasonStatusType[FailedReasonStatusType["OUT_OF_GAS"] = 3] = "OUT_OF_GAS";
    FailedReasonStatusType[FailedReasonStatusType["INVALID_INSTRUCTION"] = 4] = "INVALID_INSTRUCTION";
    FailedReasonStatusType[FailedReasonStatusType["UNDEFINED_INSTRUCTION"] = 5] = "UNDEFINED_INSTRUCTION";
    FailedReasonStatusType[FailedReasonStatusType["STACK_OVERFLOW"] = 6] = "STACK_OVERFLOW";
    FailedReasonStatusType[FailedReasonStatusType["STACK_UNDERFLOW"] = 7] = "STACK_UNDERFLOW";
    FailedReasonStatusType[FailedReasonStatusType["BAD_JUMP_DESTINATION"] = 8] = "BAD_JUMP_DESTINATION";
    FailedReasonStatusType[FailedReasonStatusType["INVALID_MEMORY_ACCESS"] = 9] = "INVALID_MEMORY_ACCESS";
    FailedReasonStatusType[FailedReasonStatusType["CALL_DEPTH_EXCEEDED"] = 10] = "CALL_DEPTH_EXCEEDED";
    FailedReasonStatusType[FailedReasonStatusType["STATIC_MODE_VIOLATION"] = 11] = "STATIC_MODE_VIOLATION";
    FailedReasonStatusType[FailedReasonStatusType["PRECOMPILE_FAILURE"] = 12] = "PRECOMPILE_FAILURE";
    FailedReasonStatusType[FailedReasonStatusType["CONTRACT_VALIDATION_FAILURE"] = 13] = "CONTRACT_VALIDATION_FAILURE";
    FailedReasonStatusType[FailedReasonStatusType["ARGUMENT_OUT_OF_RANGE"] = 14] = "ARGUMENT_OUT_OF_RANGE";
    FailedReasonStatusType[FailedReasonStatusType["WASM_UNREACHABLE_INSTRUCTION"] = 15] = "WASM_UNREACHABLE_INSTRUCTION";
    FailedReasonStatusType[FailedReasonStatusType["WASM_TRAP"] = 16] = "WASM_TRAP";
    FailedReasonStatusType[FailedReasonStatusType["INSUFFICIENT_BALANCE"] = 17] = "INSUFFICIENT_BALANCE";
    FailedReasonStatusType[FailedReasonStatusType["INTERNAL_ERROR"] = 18] = "INTERNAL_ERROR";
    FailedReasonStatusType[FailedReasonStatusType["REJECTED"] = 19] = "REJECTED";
    FailedReasonStatusType[FailedReasonStatusType["OUT_OF_MEMORY"] = 20] = "OUT_OF_MEMORY";
})(FailedReasonStatusType = exports.FailedReasonStatusType || (exports.FailedReasonStatusType = {}));
