// @lay2/pw-core/build/module/collectors/indexer-collector.js
import { SUDTCollector } from './sudt-collector';
import { Amount, AmountUnit } from '..';
import { CkbIndexer, IndexerCellToCell, ScriptType, } from '../helpers/ckb-indexer';
export class IndexerCollector extends SUDTCollector {
    constructor(apiBase) {
        super();
        this.apiBase = apiBase;
        this.indexer = new CkbIndexer(apiBase);
    }
    async getBalance(address) {
        const searchKey = {
            script: address.toLockScript().serializeJson(),
            script_type: ScriptType.lock,
            filter: {
                output_data_len_range: ['0x0', '0x1'],
            },
        };
        const cells = (await this.indexer.getCells(searchKey)).filter((cell) => cell.output.type === null);
        let balance = Amount.ZERO;
        cells.forEach((cell) => {
            const amount = new Amount(cell.output.capacity, AmountUnit.shannon);
            balance = balance.add(amount);
        });
        return balance;
    }
    async collect(address, options) {
        if (!options || !options.neededAmount) {
            throw new Error("'neededAmount' in options must be provided");
        }
        let accCapacity = Amount.ZERO;
        const terminator = (_index, cell) => {
            if (accCapacity.gte(options.neededAmount)) {
                return { stop: true, push: false };
            }
            if (cell.output_data.length / 2 - 1 > 0 || cell.output.type !== null) {
                return { stop: false, push: false };
            }
            else {
                accCapacity = accCapacity.add(new Amount(cell.output.capacity, AmountUnit.shannon));
                return { stop: false, push: true };
            }
        };
        const searchKey = {
            script: address.toLockScript().serializeJson(),
            script_type: ScriptType.lock,
            filter: {
                output_data_len_range: ['0x0', '0x1'],
            },
        };
        const cells = await this.indexer.getCells(searchKey, terminator);
        return cells.map((cell) => IndexerCellToCell(cell));
    }
    async getSUDTBalance(sudt, address) {
        const searchKey = {
            script: address.toLockScript().serializeJson(),
            script_type: ScriptType.lock,
            filter: {
                script: sudt.toTypeScript().serializeJson(),
            },
        };
        const cells = await this.indexer.getCells(searchKey);
        let balance = Amount.ZERO;
        cells.forEach((cell) => {
            const amount = Amount.fromUInt128LE(cell.output_data);
            balance = balance.add(amount);
        });
        return balance;
    }
    async collectSUDT(sudt, address, options) {
        if (!options || !options.neededAmount) {
            throw new Error("'neededAmount' in options must be provided");
        }
        const searchKey = {
            script: address.toLockScript().serializeJson(),
            script_type: ScriptType.lock,
            filter: {
                script: sudt.toTypeScript().serializeJson(),
            },
        };
        let accCapacity = Amount.ZERO;
        const terminator = (_index, cell) => {
            if (accCapacity.gte(options.neededAmount)) {
                return { stop: true, push: false };
            }
            accCapacity = accCapacity.add(Amount.fromUInt128LE(cell.output_data));
            return { stop: false, push: true };
        };
        const cells = await this.indexer.getCells(searchKey, terminator);
        return cells.map((cell) => IndexerCellToCell(cell));
    }
}
