// @lay2/pw-core/build/main/helpers/ckb-indexer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexerCellToCell = exports.CkbIndexer = exports.DefaultTerminator = exports.Order = exports.ScriptType = void 0;
const axios_1 = __importDefault(require("axios"));
const __1 = require("..");
var ScriptType;
(function (ScriptType) {
    ScriptType["type"] = "type";
    ScriptType["lock"] = "lock";
})(ScriptType = exports.ScriptType || (exports.ScriptType = {}));
var Order;
(function (Order) {
    Order["asc"] = "asc";
    Order["desc"] = "desc";
})(Order = exports.Order || (exports.Order = {}));
exports.DefaultTerminator = (_index, _cell) => {
    return { stop: false, push: true };
};
class CkbIndexer {
    constructor(ckbIndexerUrl) {
        this.ckbIndexerUrl = ckbIndexerUrl;
    }
    async request(method, params) {
        const data = {
            id: 0,
            jsonrpc: '2.0',
            method,
            params,
        };
        const res = await axios_1.default.post(this.ckbIndexerUrl, data);
        if (res.status !== 200) {
            throw new Error(`indexer request failed with HTTP code ${res.status}`);
        }
        if (res.data.error !== undefined) {
            throw new Error(`indexer request rpc failed with error: ${JSON.stringify(res.data.error)}`);
        }
        return res.data.result;
    }
    async getCells(searchKey, terminator = exports.DefaultTerminator, { sizeLimit = 0x100, order = Order.asc, } = {}) {
        const infos = [];
        let cursor = null;
        let index = 0;
        const params = [searchKey, order, `0x${sizeLimit.toString(16)}`, cursor];
        while (true) {
            const res = await this.request('get_cells', params);
            const liveCells = res.objects;
            cursor = res.lastCursor;
            for (const cell of liveCells) {
                const { stop, push } = terminator(index, cell);
                if (push) {
                    infos.push(cell);
                }
                if (stop) {
                    return infos;
                }
                index++;
            }
            if (liveCells.length < sizeLimit) {
                break;
            }
        }
        return infos;
    }
}
exports.CkbIndexer = CkbIndexer;
function IndexerCellToCell(cell) {
    return new __1.Cell(new __1.Amount(cell.output.capacity, __1.AmountUnit.shannon), __1.Script.fromRPC(cell.output.lock), __1.Script.fromRPC(cell.output.type), __1.OutPoint.fromRPC(cell.out_point), cell.output_data);
}
exports.IndexerCellToCell = IndexerCellToCell;
