// @lay2/pw-core/build/module/builders/builder.js
import { Amount, AmountUnit } from '../models';
import PWCore, { ChainID } from '..';
const FEE_BASE = 1000;
export class Builder {
    constructor(feeRate = Builder.MIN_FEE_RATE, collector = PWCore.defaultCollector, witnessArgs = PWCore.chainId === ChainID.ckb
        ? Builder.WITNESS_ARGS.RawSecp256k1
        : Builder.WITNESS_ARGS.Secp256k1) {
        this.feeRate = feeRate;
        this.collector = collector;
        this.witnessArgs = witnessArgs;
    }
    static calcFee(tx, feeRate = Builder.MIN_FEE_RATE) {
        if (feeRate < Builder.MIN_FEE_RATE) {
            feeRate = Builder.MIN_FEE_RATE;
        }
        const txSize = tx.getSize();
        const fee = (feeRate / FEE_BASE) * txSize;
        return new Amount(fee.toString(), AmountUnit.shannon);
    }
    getFee() {
        return this.fee;
    }
}
Builder.MIN_FEE_RATE = 1000;
Builder.MIN_CHANGE = new Amount('61', AmountUnit.ckb);
Builder.WITNESS_ARGS = {
    Secp256k1: {
        lock: '0x' + '0'.repeat(132),
        input_type: '',
        output_type: '',
    },
    RawSecp256k1: {
        lock: '0x' + '0'.repeat(130),
        input_type: '',
        output_type: '',
    },
    Secp256r1: {
        lock: '0x' + '0'.repeat(600),
        input_type: '',
        output_type: '',
    },
};
