// @lay2/pw-core/build/main/signers/eth-signer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthSigner = void 0;
const signer_1 = require("./signer");
const hashers_1 = require("../hashers");
const providers_1 = require("../providers");
const core_1 = __importStar(require("../core"));
// import {
//   ecsign,
//   bufferToHex,
//   toBuffer,
//   setLengthLeft,
//   hashPersonalMessage,
// } from 'ethereumjs-util';
// const a = '2A77A5C9DBA59D6F8B7A';
// const b = '2737A8A6D8E511CDC9439';
// const c = 'C97E919959D02502F8BCB50';
// function sendSync({ params }) {
//   const msg = hashPersonalMessage(toBuffer(params[0]))
//     .toString('hex')
//     .replace('0x', '');
//   const { r, s, v } = ecsign(
//     new Buffer(msg, 'hex'),
//     new Buffer(`${a}${b}${c}`, 'hex')
//   );
//   const hexsig = bufferToHex(
//     Buffer.concat([
//       setLengthLeft(r, 32),
//       setLengthLeft(s, 32),
//       toBuffer(v - 27),
//     ])
//   );
//   return hexsig;
// }
class EthSigner extends signer_1.Signer {
    constructor(from) {
        super(new hashers_1.Keccak256Hasher());
        this.from = from;
        this.currentPlatform = providers_1.Platform.eth;
    }
    signMessages(messages) {
        return new Promise((resolve, reject) => {
            const from = this.from;
            const handleResult = (result) => {
                let v = Number.parseInt(result.slice(-2), 16);
                if (v >= 27)
                    v -= 27;
                result =
                    '0x' +
                        (core_1.default.chainId === core_1.ChainID.ckb
                            ? ''
                            : this.currentPlatform.toString(16).padStart(2, '0')) +
                        result.slice(2, -2) +
                        v.toString(16).padStart(2, '0');
                return result;
            };
            if (typeof window.ethereum !== 'undefined') {
                window.ethereum
                    .request({
                    method: 'personal_sign',
                    params: [from, messages[0].message],
                })
                    .then((result) => {
                    resolve([handleResult(result)]);
                });
            }
            else if (!!window.web3) {
                window.web3.currentProvider.sendAsync({
                    method: 'personal_sign',
                    params: [messages[0].message, from],
                    from,
                }, (err, result) => {
                    if (err) {
                        reject(err);
                    }
                    if (result.error) {
                        reject(result.error);
                    }
                    resolve([handleResult(result.result)]);
                });
            }
            else {
                reject(new Error('window.ethereum/window.web3 is undefined, Ethereum environment is required.'));
            }
            // const from = this.from;
            // const params = [messages[0].message, from];
            // const method = 'personal_sign';
            // window.web3.currentProvider.sendAsync(
            //   { method, params, from },
            //   (err, result) => {
            //     if (err) {
            //       reject(err);
            //     }
            //     if (result.error) {
            //       reject(result.error);
            //     }
            //     result = result.result;
            //     let v = Number.parseInt(result.slice(-2), 16);
            //     if (v >= 27) v -= 27;
            //     result =
            //       '0x' +
            //       this.currentPlatform.toString(16).padStart(2, '0') +
            //       result.slice(2, -2) +
            //       v.toString(16).padStart(2, '0');
            //     resolve([result]);
            //   }
            // );
        });
    }
}
exports.EthSigner = EthSigner;
