// @lay2/pw-core/build/main/signers/signer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Signer = void 0;
const hashers_1 = require("../hashers");
const ckb_js_toolkit_1 = require("../ckb-js-toolkit");
const core_1 = require("../ckb-lumos/core");
class Signer {
    constructor(hasher) {
        this.hasher = hasher;
    }
    async sign(tx) {
        const messages = this.toMessages(tx);
        const witnesses = await this.signMessages(messages);
        for (let i = 0; i < messages.length; i++) {
            const { index } = messages[i];
            if (index < tx.witnessArgs.length &&
                typeof tx.witnessArgs[index] !== 'string') {
                witnesses[i] = new ckb_js_toolkit_1.Reader(core_1.SerializeWitnessArgs(ckb_js_toolkit_1.normalizers.NormalizeWitnessArgs(Object.assign(Object.assign({}, tx.witnessArgs[index]), { lock: witnesses[i] })))).serializeJson();
            }
        }
        tx = FillSignedWitnesses(tx, messages, witnesses);
        return tx;
    }
    toMessages(tx) {
        tx.validate();
        if (tx.raw.inputs.length !== tx.raw.inputCells.length) {
            throw new Error('Input number does not match!');
        }
        const txHash = new hashers_1.Blake2bHasher().hash(new ckb_js_toolkit_1.Reader(core_1.SerializeRawTransaction(ckb_js_toolkit_1.normalizers.NormalizeRawTransaction(ckb_js_toolkit_1.transformers.TransformRawTransaction(tx.raw)))));
        const messages = [];
        const used = tx.raw.inputs.map((_input) => false);
        for (let i = 0; i < tx.raw.inputs.length; i++) {
            if (used[i]) {
                continue;
            }
            if (i >= tx.witnesses.length) {
                throw new Error(`Input ${i} starts a new script group, but witness is missing!`);
            }
            used[i] = true;
            this.hasher.update(txHash);
            const firstWitness = new ckb_js_toolkit_1.Reader(tx.witnesses[i]);
            this.hasher.update(serializeBigInt(firstWitness.length()));
            this.hasher.update(firstWitness);
            for (let j = i + 1; j < tx.raw.inputs.length && j < tx.witnesses.length; j++) {
                if (tx.raw.inputCells[i].lock.sameWith(tx.raw.inputCells[j].lock)) {
                    used[j] = true;
                    const currentWitness = new ckb_js_toolkit_1.Reader(tx.witnesses[j]);
                    this.hasher.update(serializeBigInt(currentWitness.length()));
                    this.hasher.update(currentWitness);
                }
            }
            messages.push({
                index: i,
                message: this.hasher.digest().serializeJson(),
                lock: tx.raw.inputCells[i].lock,
            });
            this.hasher.reset();
        }
        return messages;
    }
}
exports.Signer = Signer;
function FillSignedWitnesses(tx, messages, witnesses) {
    if (messages.length !== witnesses.length) {
        throw new Error('Invalid number of witnesses!');
    }
    for (let i = 0; i < messages.length; i++) {
        tx.witnesses[messages[i].index] = witnesses[i];
    }
    return tx;
}
function serializeBigInt(i) {
    const view = new DataView(new ArrayBuffer(8));
    view.setUint32(0, i, true);
    return view.buffer;
}
